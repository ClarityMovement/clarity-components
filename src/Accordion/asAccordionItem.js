import { compose, withProps, withHandlers } from 'recompose';
import { embedHandler } from 'react-render-counter/hocs';

const styles = {
  narrow: {
    default: {
      flex: 1,
      width: '100%',
      overflow: 'hidden',
      transition: 'all 0.2s ease-in',
      opacity: 1,
    },
    collapsed: {
      flex: 0,
      transition: 'all 0.2s ease-out',
      opacity: 0,
    },
    expanded: {
      flex: 2,
      transition: 'all 0.2s ease-out',
      opacity: 1,
    },
  },
  wide: {
    default: {
      width: '100%',
    },
    collapsed: {
      margin: 0,
      transition: 'all 0.2s ease-out',
      width: '100%',
    },
    expanded: {
      margin: '50px 0',
      width: 'calc(100% + 40px)',
      transition: 'all 0.2s ease-in',
    },
  },
};

const contentStyles = {
  default: {
    overflow: 'hidden',
  },
  collapsed: {
    transition: 'all 0.2s ease-in',
    padding: 0,
    maxHeight: 0,
    opacity: 0,
  },
  expanded: {
    transition: 'all 0.2s ease-out',
    maxHeight: '100vh',
    opacity: 1,
  },
};

const stylesMapper = ({ wide, expanded, collapsed, style }) => {
  const rootStyles = styles[wide ? 'wide' : 'narrow'];
  const rootStyle = { ...rootStyles.default };
  const contentStyle = {
    ...contentStyles.default,
    ...contentStyles[expanded ? 'expanded' : 'collapsed'],
  };

  if (expanded) {
    Object.assign(rootStyle, rootStyles.expanded);
  }
  if (collapsed) {
    Object.assign(rootStyle, rootStyles.collapsed);
  }

  return {
    style: { ...rootStyle, ...style },
    contentStyle,
  };
};

const propsMapper = ({ activeKey, itemKey }) => ({
  expanded: activeKey === itemKey,
  collapsed: activeKey != null && activeKey !== itemKey,
});

export const handlers = {
  onExpand: ({ onActivate, itemKey }) => event => onActivate(event, itemKey),
  onCollapse: ({ onActivate }) => event => onActivate(event, null),
};

const asAccordionItem = compose(
  embedHandler('onActivate', 'onToggle'),
  withHandlers(handlers),
  withProps(propsMapper),
  withProps(stylesMapper),
);

const uniquify = uniqueHoc => (Component) => {
  const hocs = Component.hocs || [];
  if (hocs.find(hoc => hoc === uniqueHoc)) {
    return Component;
  }
  const EnhancedComponent = uniqueHoc(Component);
  EnhancedComponent.hocs = hocs.concat(uniqueHoc);

  return EnhancedComponent;
};

export default uniquify(asAccordionItem);
