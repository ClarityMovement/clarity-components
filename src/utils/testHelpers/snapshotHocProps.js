import React from 'react';

import enzymeToJson from 'enzyme-to-json';
import { shallowWithUntil as shallow } from './until';

const snapshotPropsMapperHoc = (hoc, inputProps) => {
  // This Component can't render deeply, must work with shallow renderer.
  const PropsDisplayer = outputProps => (
    <div name="props-displayer">
      <input-props {...inputProps} />
      <output-props {...outputProps} />
    </div>
  );
  const NewComponent = hoc(PropsDisplayer);
  const wrapper = shallow(<NewComponent {...inputProps} />).until(PropsDisplayer);
  expect(enzymeToJson(wrapper)).toMatchSnapshot();
};

export default snapshotPropsMapperHoc;
