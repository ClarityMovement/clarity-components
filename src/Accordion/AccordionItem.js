import React from 'react';
import T from 'prop-types';
import { Card, CardText, ListItem } from 'material-ui';
import CloseIcon from 'material-ui/svg-icons/navigation/close';
import asAccordionItem from './asAccordionItem';
import { stylePropType } from './constants';

export const AccordionItemRenderer = ({
  style,
  contentStyle,
  wide,
  expanded,
  onExpand,
  onCollapse,
  title,
  children,
}) => (
  <Card style={style}>
    <ListItem
      primaryText={title}
      onClick={wide && expanded ? onCollapse : onExpand}
      leftIcon={!wide && expanded ? <CloseIcon onClick={onCollapse} /> : null}
    />
    <CardText style={contentStyle}>
      {children}
    </CardText>
  </Card>
);

AccordionItemRenderer.propTypes = {
  style: stylePropType,
  contentStyle: stylePropType,
  wide: T.bool,
  expanded: T.bool,
  onExpand: T.func,
  onCollapse: T.func,
  title: T.string,
  children: T.node,
};

AccordionItemRenderer.defaultProps = {
  style: {},
  contentStyle: {},
  wide: undefined,
  expanded: undefined,
  onExpand: undefined,
  onCollapse: undefined,
  title: undefined,
  children: undefined,
};

export default asAccordionItem(AccordionItemRenderer);
