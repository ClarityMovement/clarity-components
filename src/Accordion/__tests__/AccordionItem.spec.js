/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import enzymeToJson from 'enzyme-to-json';
import snapshotHocProps from 'react-render-counter/utils/testHelpers/snapshotHocProps';
import AccordionItem, { AccordionItemRenderer } from '../AccordionItem';
import asAccordionItem from '../asAccordionItem';

describe('<AccordionItem {...props} />: elements tree', () => {
  let props;
  beforeEach(() => {
    props = { ...AccordionItemRenderer.defaultProps };
  });

  afterEach(() => {
    snapshotHocProps(asAccordionItem, props);

    const wrapper = shallow(<AccordionItem {...props} />).until(AccordionItemRenderer);
    expect(enzymeToJson(wrapper)).toMatchSnapshot();
  });

  describe('when device is narrow', () => {
    test('when item is expanded', () => {
      props = {
        ...props,
        itemKey: 1,
        activeKey: 1,
      };
    });

    test('when item is collapsed', () => {
      props = {
        ...props,
        itemKey: 1,
        activeKey: 0,
      };
    });

    test('when animation is off', () => {
      props = {
        ...props,
        animation: false,
      };
    });
  });

  describe('when device is wide', () => {
    beforeEach(() => {
      props.wide = true;
    });

    test('when item is expanded and the first', () => {
      props = {
        ...props,
        isFirst: true,
        itemKey: 1,
        activeKey: 1,
      };
    });

    test('when item is collapsed and the last', () => {
      props = {
        ...props,
        isLast: false,
        itemKey: 1,
        activeKey: 0,
      };
    });
  });
});
