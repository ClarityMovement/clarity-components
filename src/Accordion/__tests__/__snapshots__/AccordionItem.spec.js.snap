// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AccordionItem {...props} />: elements tree when device is narrow when item is collapsed 1`] = `
<div
  name="props-displayer"
>
  <input-props
    activeKey={0}
    contentStyle={Object {}}
    itemKey={1}
    style={Object {}}
  />
  <output-props
    activeKey={0}
    collapsed={true}
    contentStyle={
      Object {
        "maxHeight": 0,
        "opacity": 0,
        "overflow": "hidden",
        "padding": 0,
        "transition": "all 0.2s ease-in",
      }
    }
    expanded={false}
    itemKey={1}
    onActivate={[Function]}
    onCollapse={[Function]}
    onExpand={[Function]}
    style={
      Object {
        "flex": 0,
        "opacity": 0,
        "overflow": "hidden",
        "transition": "all 0.2s ease-out",
        "width": "100%",
      }
    }
  />
</div>
`;

exports[`<AccordionItem {...props} />: elements tree when device is narrow when item is collapsed 2`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "flex": 0,
      "opacity": 0,
      "overflow": "hidden",
      "transition": "all 0.2s ease-out",
      "width": "100%",
    }
  }
>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    leftIcon={null}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryTogglesNestedList={false}
    secondaryTextLines={1}
  />
  <CardText
    style={
      Object {
        "maxHeight": 0,
        "opacity": 0,
        "overflow": "hidden",
        "padding": 0,
        "transition": "all 0.2s ease-in",
      }
    }
  />
</Card>
`;

exports[`<AccordionItem {...props} />: elements tree when device is narrow when item is collapsed 3`] = `
<div
  name="props-displayer"
>
  <input-props
    activeKey={0}
    contentStyle={Object {}}
    itemKey={1}
    style={Object {}}
    wide={true}
  />
  <output-props
    activeKey={0}
    collapsed={true}
    contentStyle={
      Object {
        "maxHeight": 0,
        "opacity": 0,
        "overflow": "hidden",
        "padding": 0,
        "transition": "all 0.2s ease-in",
      }
    }
    expanded={false}
    itemKey={1}
    onActivate={[Function]}
    onCollapse={[Function]}
    onExpand={[Function]}
    style={
      Object {
        "margin": 0,
        "transition": "all 0.2s ease-out",
        "width": "100%",
      }
    }
    wide={true}
  />
</div>
`;

exports[`<AccordionItem {...props} />: elements tree when device is narrow when item is collapsed 4`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "margin": 0,
      "transition": "all 0.2s ease-out",
      "width": "100%",
    }
  }
>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    leftIcon={null}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryTogglesNestedList={false}
    secondaryTextLines={1}
  />
  <CardText
    style={
      Object {
        "maxHeight": 0,
        "opacity": 0,
        "overflow": "hidden",
        "padding": 0,
        "transition": "all 0.2s ease-in",
      }
    }
  />
</Card>
`;

exports[`<AccordionItem {...props} />: elements tree when device is narrow when item is expanded 1`] = `
<div
  name="props-displayer"
>
  <input-props
    activeKey={1}
    contentStyle={Object {}}
    itemKey={1}
    style={Object {}}
  />
  <output-props
    activeKey={1}
    collapsed={false}
    contentStyle={
      Object {
        "maxHeight": "100vh",
        "opacity": 1,
        "overflow": "hidden",
        "transition": "all 0.2s ease-out",
      }
    }
    expanded={true}
    itemKey={1}
    onActivate={[Function]}
    onCollapse={[Function]}
    onExpand={[Function]}
    style={
      Object {
        "flex": 2,
        "opacity": 1,
        "overflow": "hidden",
        "transition": "all 0.2s ease-out",
        "width": "100%",
      }
    }
  />
</div>
`;

exports[`<AccordionItem {...props} />: elements tree when device is narrow when item is expanded 2`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "flex": 2,
      "opacity": 1,
      "overflow": "hidden",
      "transition": "all 0.2s ease-out",
      "width": "100%",
    }
  }
>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    leftIcon={
      <NavigationClose
        onTouchTap={[Function]}
      />
    }
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryTogglesNestedList={false}
    secondaryTextLines={1}
  />
  <CardText
    style={
      Object {
        "maxHeight": "100vh",
        "opacity": 1,
        "overflow": "hidden",
        "transition": "all 0.2s ease-out",
      }
    }
  />
</Card>
`;

exports[`<AccordionItem {...props} />: elements tree when device is narrow when item is expanded 3`] = `
<div
  name="props-displayer"
>
  <input-props
    activeKey={1}
    contentStyle={Object {}}
    itemKey={1}
    style={Object {}}
    wide={true}
  />
  <output-props
    activeKey={1}
    collapsed={false}
    contentStyle={
      Object {
        "maxHeight": "100vh",
        "opacity": 1,
        "overflow": "hidden",
        "transition": "all 0.2s ease-out",
      }
    }
    expanded={true}
    itemKey={1}
    onActivate={[Function]}
    onCollapse={[Function]}
    onExpand={[Function]}
    style={
      Object {
        "margin": "50px 0",
        "transition": "all 0.2s ease-in",
        "width": "calc(100% + 40px)",
      }
    }
    wide={true}
  />
</div>
`;

exports[`<AccordionItem {...props} />: elements tree when device is narrow when item is expanded 4`] = `
<Card
  expandable={false}
  expanded={null}
  initiallyExpanded={false}
  style={
    Object {
      "margin": "50px 0",
      "transition": "all 0.2s ease-in",
      "width": "calc(100% + 40px)",
    }
  }
>
  <ListItem
    autoGenerateNestedIndicator={true}
    containerElement="span"
    disableKeyboardFocus={false}
    disabled={false}
    initiallyOpen={false}
    insetChildren={false}
    leftIcon={null}
    nestedItems={Array []}
    nestedLevel={0}
    onKeyboardFocus={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onNestedListToggle={[Function]}
    onTouchEnd={[Function]}
    onTouchStart={[Function]}
    onTouchTap={[Function]}
    open={null}
    primaryTogglesNestedList={false}
    secondaryTextLines={1}
  />
  <CardText
    style={
      Object {
        "maxHeight": "100vh",
        "opacity": 1,
        "overflow": "hidden",
        "transition": "all 0.2s ease-out",
      }
    }
  />
</Card>
`;
